script: Handle broken image icon in `HTMLImageElement`

mrobinson opened the pull request:
Instead of having the `ImageCache` return the broken image icon for
failed loads, have `HTMLImageElement` explicitly request it. This means
that the image is loaded on demand (reducing the usage of resources) and
also simplifying the interface of the `ImageCache` greatly.

In addition, the display of the broken image icon is improved, more
closely matching other browsers. A test for this display (which was
falsely passing before) is updated to reflect the new display of the
broken image icon.

Testing: There is a Servo-specific test for this change. Some WPT tests
start to fail as well. Before these were not properly loading the broken
image icon so they were failing before, just in a hidden way.

mrobinson committed:
script: Handle broken image icon in `HTMLImageElement`

Instead of having the `ImageCache` return the broken image icon for
failed loads, have `HTMLImageElement` explicitly request it. This means
that the image is loaded on demand (reducing the usage of resources) and
also simplifying the interface of the `ImageCache` greatly.

In addition, the display of the broken image icon is improved, more
closely matching other browsers. A test for this display (which was
falsely passing before) is updated to reflect the new display of the
broken image icon.

Co-authored-by: Mukilan Thiyagarajan <mukilan@igalia.com>
Signed-off-by: Martin Robinson <mrobinson@igalia.com>

jdm reviewed (approved):
Excellent API improvement!

mrobinson committed:
Adjust sizing logic for the broken image icon

- The aspect ratio of the broken image icon should not affect the
  preferred_aspect_ratio of images.
- Simplify some of the sizing logic.
- Make a calculate_fragment_rect and add an early return for images.
- Have `naturalWidth` and `naturalHeight` return 0 if the image is
  broken.

Some tests are marked as failing, because the broken image icon wasn't
showing up for them in the past, but now it is.

Signed-off-by: Martin Robinson <mrobinson@igalia.com>
